rpcs3/rpcs3qt/settings_dialog.cpp

```diff
@@ -1042,6 +1039,43 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
 	m_emu_settings->EnhanceComboBox(ui->vulkansched, emu_settings_type::VulkanAsyncSchedulerDriver);
 	SubscribeTooltip(ui->gb_vulkansched, tooltips.settings.vulkan_async_scheduler);
 
+	// Custom control that simplifies operation of two independent variables. Can probably be done better but this works.
+	ui->zcullPrecisionMode->addItems({"Precise (Default)", "Approximate (Fast)", "Relaxed (Fastest)"});
+	const auto is_relaxed_zcull = m_emu_settings->GetSetting(emu_settings_type::RelaxedZCULL) == "true";
+	const auto is_precise_zcull = m_emu_settings->GetSetting(emu_settings_type::PreciseZCULL) == "true";
+	if (is_relaxed_zcull)
+	{
+		ui->zcullPrecisionMode->setCurrentIndex(2);
+	}
+	else if (is_precise_zcull)
+	{
+		ui->zcullPrecisionMode->setCurrentIndex(0);
+	}
+	else
+	{
+		ui->zcullPrecisionMode->setCurrentIndex(1);
+	}
+	connect(ui->zcullPrecisionMode, QOverload<int>::of(&QComboBox::currentIndexChanged), [this](int index)
+	{
+		bool relaxed = false, precise = false;
+
+		switch (index)
+		{
+		case 0:
+			precise = true; break;
+		case 1:
+			break;
+		case 2:
+			relaxed = true; break;
+		default:
+			fmt::throw_exception("Unexpected selection");
+		}
```

Comment:
So if you pass userData you need to decode it here

```suggestion
		switch (ui->zcullPrecisionMode->itemData(index).toInt())
		{
		case 0:
			break;
		case static_cast<int>(emu_settings_type::PreciseZCULL):
			precise = true; break;
		case static_cast<int>(emu_settings_type::RelaxedZCULL):
			relaxed = true; break;
		default:
			fmt::throw_exception("Unexpected selection");
		}
```

Reply:
Why bother? The options are abstract and not directly tied into the actual options they control that way.
If passing user data, it would just match 0-1-2 anyway. Feels like needless overdesign to pass in the same values as currentIndex.

Reply:
Why should it be 1 and 2 ? that makes no sense at all.
And you didn't read the part about sorting or changing the order

Reply:
Its not tied to the enum values at all, that is a coincidence in this specific case. This is actually a scale from most accurate to most performant - I can replace it with a slider if it makes more sense.
It just so happens that the options that affect each setting can be mapped this way, but this is arbitrary, we could just as well have 10 options controlled by this abstraction.
I could also just remove the whole thing and have users go through another checkbox, but I doubt regular users would understand when to use the options without looking up a wiki. A simple scale is much easier to understand.

Reply:
I think you're missing the point here

Reply:
We're clearly not on the same page. The switch is what is causing confusion here, it is not meant to toggle things on/off. Otherwise if I end up with a third control parameter, now I will need to pass bitmasks as payload, increasing the complexity. One level can control an arbitrary number of switches, not just 1 or 2 as it is right now. I'm trying to keep my options open here, I need to add more stuff later.

Reply:
It is much more obvious if I add a third parameter X which is on for relaxed and precise but not for approximate.

Reply:
what i mean is -> you now have hardcoded values 0 1 and 2 which only ever match the visual order in the dropdown and are also not very easy to understand (keyword magic numbers).
If you use the itemData, you make it foolproof by shifting the logic to the model instead of the item order.
You can then also identify it much easier while reading the code, because the switch is listing the enum values instead of the indices.
As another resulting extra you can then also use setCurrentIndex(findData(enum)) instead of setCurrentIndex(2), which removes another set of magic numbers.

The 0 will still be a magic number, but since there's no setting for that, there's no way around it.

Reply:
Ok, I'll add an enum, I suppose GUI isn't really performance critical. My argument was that we already have an implicit model here - a non-repeating incrementing set that corresponds 1:1 with displayed text. The extra values passed are just going to be the same thing, and since the set is also non-repeating, it is technically unnecessary. I suppose it can harm readability somewhat though.

Reply:
Fixed.